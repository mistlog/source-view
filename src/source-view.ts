import { isTSTypeAliasDeclaration, isExportNamedDeclaration, isExpressionStatement, Node, File, isImportDeclaration, isEmptyStatement, Comment } from "@babel/types";
import generate from "@babel/generator";
import { parse } from "@babel/parser";
import * as prettier from "prettier";

export function CodeToMarkdown(code: string)
{
    return new SourceView(code).ToMarkdown();
}

interface IParagraph
{
    text: string;
    type: string;
    start: number;
}

class SourceView
{
    m_File: File;
    m_Document: Array<IParagraph>;
    m_CommentSet: Set<number>;

    constructor(code: string)
    {
        const with_semi = this.CodeWithSemi(code);
        this.m_File = parse(with_semi, { sourceType: "module", plugins: ["typescript", "jsx", "optionalChaining"] });
        this.m_Document = [];
        this.m_CommentSet = new Set<number>();
    }

    /**
     * add semi after interface declaration to avoid parse error:
     * eg.
     *      interface foo{}
     *      (add ; before it)<Foo> + function...
     */
    CodeWithSemi(raw: string)
    {
        const code = prettier.format(raw, {
            parser: "typescript",
            semi: false
        });
        return code;
    }

    ToMarkdown(): string
    {
        //
        const body = this.m_File.program.body;
        body.forEach(each => this.NodeToMarkdown(each));

        //
        this.m_Document.sort((a, b) => a.start - b.start);
        const markdown = this.m_Document.map(paragraph =>
        {
            if (paragraph.type === "code")
            {
                return ["```typescript", paragraph.text, "```"].join("\r\n");
            }
            return paragraph.text;
        }).join("\r\n\r\n");
        return markdown;
    }

    AddToDocument(text: string, type: string, start: number)
    {
        if (this.m_CommentSet.has(start))
        {
            return;
        }

        this.m_CommentSet.add(start);
        this.m_Document.push({
            text: text.trim(),
            type,
            start
        });
    }

    NodeToMarkdown(node: Node)
    {
        //
        let leading = node.leadingComments;
        let trailing = node.trailingComments;

        //
        this.AddCommentsToDocument(leading);

        //
        node.leadingComments = [];
        node.trailingComments = [];

        if (!isImportDeclaration(node) && !isEmptyStatement(node))
        {
            /**
            * trailing comments were moved to expression after add semi by babel
            */
            if (isExpressionStatement(node) && !trailing)
            {
                trailing = node.expression.trailingComments;
                node.expression.trailingComments = [];
            }
            else if (isExportNamedDeclaration(node) && isTSTypeAliasDeclaration(node.declaration) && !trailing)
            {
                trailing = node.declaration.typeAnnotation.trailingComments;
                node.declaration.typeAnnotation.trailingComments = [];
            }

            const code = this.NodeToString(node);
            this.AddToDocument(code, "code", node.start);
        }

        //
        this.AddCommentsToDocument(trailing);
    }

    AddCommentsToDocument(comments: ReadonlyArray<Comment>)
    {
        comments?.forEach(each => this.AddToDocument(each.value, "markdown", each.start));
    }

    NodeToString(node: Node)
    {
        // minify it to handle new line around comment
        const raw = generate(node, {
            shouldPrintComment: comment => comment === "@ts-ignore" ? false : true,
            minified: true
            // add new line around comment
        }).code.replace(/\*\//g, "*/\r\n").
            replace(/\/\*/g, "\r\n\r\n/*").
            replace(/\/\//g, "\r\n\r\n//");

        const code = prettier.format(raw, {
            parser: "typescript",
            printWidth: 200,
            tabWidth: 4
        });
        return code;
    }
}