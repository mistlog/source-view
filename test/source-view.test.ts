import { readFileSync, writeFileSync } from "fs";
import { CodeToMarkdown } from "../src/source-view";

describe("source view", () =>
{
    test("code and markdown", () =>
    {
        SnapshotTest("code-and-markdown");
    })

    test("exclude import", () =>
    {
        SnapshotTest("exclude-import");
    })

    test("exclude empty statement", () =>
    {
        SnapshotTest("exclude-empty-statement");
    })

    test("remove @ts-ignore", () =>
    {
        SnapshotTest("remove-tsignore");
    })

    test("handle comment", () =>
    {
        SnapshotTest("comment-order");
    })

    test("handle semi", () =>
    {
        SnapshotTest("handle-semi");
    })

    test("expression statement", () =>
    {
        SnapshotTest("expression-statement");
    })

    test("export type", () =>
    {
        SnapshotTest("export-type");
    })

    test("optional chaining", () =>
    {
        SnapshotTest("optional-chaining");
    })
})

/**
 * utility
 */

function SnapshotTest(name: string)
{

    const code = ReadInput(name);
    const markdown = CodeToMarkdown(code);
    WriteOutput(name, markdown);
    expect(markdown).toMatchSnapshot();
}

function WriteOutput(name: string, markdown: string)
{
    writeFileSync(`${__dirname}/common/output/${name}.md`, markdown);
}

function ReadInput(name: string)
{
    return readFileSync(`${__dirname}/common/input/${name}.tsx`, "utf8")
}