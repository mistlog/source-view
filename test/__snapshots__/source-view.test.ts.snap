// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`source view code and markdown 1`] = `
"# Code and markdown

\`\`\`typescript
class Test {}
\`\`\`

\`\`\`typescript
function foo() {
    console.log(\\"hi\\");
}
\`\`\`"
`;

exports[`source view exclude empty statement 1`] = `
"\`\`\`typescript
export interface foo {}
\`\`\`

\`\`\`typescript
function foo3() {}
\`\`\`"
`;

exports[`source view exclude import 1`] = `
"# test

\`\`\`typescript
function foo2() {}
\`\`\`"
`;

exports[`source view export type 1`] = `
"\`\`\`typescript
export type ITraverseMethodCallback = (methods: Array<ClassMethod>, class_name: string) => void;
\`\`\`

# Implementation

##  DSL

\`\`\`typescript
<ExportClassCode /> +
    function constructor(this: ExportClassCode, node: ExportNamedDeclaration, path?: NodePath<ExportNamedDeclaration>) {
        this.m_Code = node;
        this.m_Path = path;
    };
\`\`\`"
`;

exports[`source view expression statement 1`] = `
"# Class

All exported class would be recorded as \`\`\`ExportClassCode\`\`\`,

\`\`\`typescript
export class ExportClassCode {
    m_Code: ExportNamedDeclaration;
    m_Path: NodePath<ExportNamedDeclaration>;
    get m_Class() {
        return this.m_Code.declaration as ClassDeclaration;
    }
    get m_Members() {
        return this.m_Class.body.body;
    }
    get m_Name() {
        return this.m_Class.id.name;
    }
}
\`\`\`

Then, we will restore class methods and add them to class. See document of method.

\`\`\`typescript
<ExportClassCode /> +
    function AddMember(this: ExportClassCode, member: ClassMethod) {
        this.m_Members.push(member);
    };
\`\`\`

# Trivial

\`\`\`typescript
<ExportClassCode /> +
    function constructor(this: ExportClassCode, node: ExportNamedDeclaration, path?: NodePath<ExportNamedDeclaration>) {
        this.m_Code = node;
        this.m_Path = path;
    };
\`\`\`"
`;

exports[`source view handle comment 1`] = `
"leading comment of func1

\`\`\`typescript
function func1() {}
\`\`\`

trailing comment of func1 and leading comment of func2

\`\`\`typescript
function func2() {}
\`\`\`

trailing comment of func2"
`;

exports[`source view handle semi 1`] = `
"\`\`\`typescript
class Foo {}
\`\`\`

\`\`\`typescript
export interface Test {}
\`\`\`

\`\`\`typescript
<Foo /> + function constructor() {};
\`\`\`"
`;

exports[`source view remove @ts-ignore 1`] = `
"\`\`\`typescript
function ignore() {
    //other comment
    console.log(\\"hi\\");
}
\`\`\`"
`;
